### PR TITLE
refactor/better_sleep_mode

### PR DESCRIPTION
**Pull request by JarbasAl**
builds on #2 

the sleep mode becomes much snappier and accurate now:
- the wake up detection recorded a full utterance and wake up detection was performed on it in the STT stage, this recording could be very long and made detections difficult
- no longer needs to break out of the wake word loop to process the wake up command, it is already checking for hotwords so the wake up detection was moved there

### CODE REVIEW

File: mycroft/client/speech/listener.py [hunk 1]
```diff
@@ -310,7 +300,7 @@ def create_hotword_engines(self):
                 sound = data.get("sound")
                 utterance = data.get("utterance")
                 listen = data.get("listen", False)
-                wakeup = data.get("wake_up", False)
+                wakeup = data.get("wakeup", False)
```
Comment (JarbasAl) [inline mycroft/client/speech/listener.py:303]: bug introduced in #2

File: mycroft/client/speech/mic.py [hunk 2]
```diff
@@ -780,9 +801,9 @@ def listen(self, source, stream):
         LOG.debug("Waiting for wake word...")
         ww_data, lang = self._wait_until_wake_word(source, sec_per_buffer)
 
-        if ww_data.stopped:
-            # If the waiting returned from a stop signal
-            return
+        if ww_data.stopped or self.loop.state.sleeping:
+            # If the waiting returned from a stop signal or sleep mode is active
+            return None, lang
```
Comment (JarbasAl) [inline mycroft/client/speech/mic.py:806]: bug introduced in #2 or maybe earlier in the multilingual support commit

File: mycroft/client/speech/mic.py [hunk 3]
```diff
@@ -359,6 +360,8 @@ def __init__(self, loop, watchdog=None):
         # Signal statuses
         self._stop_signaled = False
         self._listen_triggered = False
+        self._waiting_for_wakeup = False
```
Comment (NeonDaniel) [inline mycroft/client/speech/mic.py:363]: Should this be set to loop.state.sleeping?

**Review by NeonDaniel — COMMENTED** [review]
I'm probably not seeing something, but in `mic.py`, is `self._waiting_for_wakeup` not always in the same state as `loop.state.sleeping`?

File: mycroft/client/speech/mic.py (hunk 3 above)
Comment (JarbasAl) [reply mycroft/client/speech/mic.py:363]: no, this is a flag for the wake word detection when in sleep mode, it should only be set to True when "hey mycroft" is heard in sleep mode. This basically starts a 5 second time window when "wake up" is a valid detection

Comment (JarbasAl) [reply mycroft/client/speech/mic.py:363]: > I'm probably not seeing something, but in `mic.py`, is `self._waiting_for_wakeup` not always in the same state as `loop.state.sleeping`?

you enter in sleep mode and self._waiting_for_wakeup is False, it only becomes True (for 5 seconds) once a valid wake word is heard, after 5 seconds goes back to False.

**Review by NeonDaniel — APPROVED** [review]

**Pull request merged**